web/service/src/main/java/org/mobilitydata/gtfsvalidator/web/service/util/StorageHelper.java

```diff
@@ -0,0 +1,166 @@
+      Blob blob = storage.get(jobBlobId);
+      var json = new String(blob.getContent());
+      logger.info("Loading job metadata: " + json);
+
+      var objectMapper = new ObjectMapper();
+      JobMetadata jobMetaData = objectMapper.readValue(json, JobMetadata.class);
+      return jobMetaData.getCountryCode();
+    } catch (Exception exc) {
+      logger.error("Error could not load remote file, using default country code", exc);
+      return "";
+    }
+  }
+
+  /**
+   * Saves a file from a URL to GCS at a job-specific path.
+   *
+   * @param jobId
+   * @param url
+   * @throws Exception
+   */
+  public void saveJobFileFromUrl(String jobId, String url) throws Exception {
+    // Read file into memory
+    var urlInputStream = new BufferedInputStream(new URL(url).openStream());
+
+    // Upload to GCS
+    var blobId = BlobId.of(USER_UPLOAD_BUCKET_NAME, jobId + "/" + jobId + ".zip");
+    var mimeType = "application/zip";
+    var blobInfo = BlobInfo.newBuilder(blobId).setContentType(mimeType).build();
+    var fileBytes = urlInputStream.readAllBytes();
+    storage.create(blobInfo, fileBytes);
+  }
+
+  /** Generates a job-specific signed URL for uploading a file to GCS. */
+  public URL generateUniqueUploadUrl(String jobId) {
+    var blobInfo =
+        BlobInfo.newBuilder(BlobId.of(USER_UPLOAD_BUCKET_NAME, jobId + "/" + FILE_NAME)).build();
+
+    // Generate Signed URL
+    Map<String, String> extensionHeaders = new HashMap<>();
+    extensionHeaders.put("Content-Type", "application/octet-stream");
+
+    URL url =
+        storage.signUrl(
+            blobInfo,
+            15,
+            TimeUnit.MINUTES,
+            Storage.SignUrlOption.httpMethod(HttpMethod.PUT),
+            Storage.SignUrlOption.withExtHeaders(extensionHeaders),
+            Storage.SignUrlOption.withV4Signature());
+    return url;
+  }
+
+  /**
+   * Copies the uploaded feed file from GCS to a local temp directory and returns the file.
+   *
+   * @param jobId
+   * @param fileName
+   * @return
+   * @throws IOException
+   */
+  public File createTempFile(String jobId, String fileName) throws IOException {
+    var tempDir = Files.createTempDirectory(StorageHelper.TEMP_FOLDER_NAME).toFile();
+
+    var inputResource =
+        applicationContext.getResource("gs://" + USER_UPLOAD_BUCKET_NAME + "/" + fileName);
+    var inputFile = inputResource.getInputStream();
+
+    var tempFile = File.createTempFile(jobId, ".zip", tempDir);
+    var output = new FileOutputStream(tempFile);
+    inputFile.transferTo(output);
+    return tempFile;
+  }
+
+  /**
+   * Uploads the validation report files to GCS.
+   *
+   * @param jobId
+   * @param outputPath
+   * @throws IOException
+   */
+  public void uploadFilesToStorage(String jobId, File outputPath) throws IOException {
+    var directoryListing = outputPath.listFiles();
+    if (directoryListing != null) {
+      for (var reportFile : directoryListing) {
```

Comment:
What would happen if there is a sub-directory here?

Reply:
@bdferris-v2 I'm not sure.  Would look for guidance on how to handle that, I'm not sure what `listFiles` returns for subdirectories and/or how we would prefer to handle that (preserve, flatten, error?)

/cc @KClough

Reply:
Maybe just add a defensive `if (reportFile.isDirectory()) { continue; }` just to be safe?